iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+    for (auto opPair : opPairs) {
+      auto &lhsOp = std::get<0>(opPair);
+      auto &rhsOp = std::get<1>(opPair);
+      if (!OperationEquivalence::isEquivalentTo(
+              &lhsOp, &rhsOp, OperationEquivalence::IgnoreOperands)) {
+        return false;
+      }
+
+      // We want to check the operand _types_, but don't care if the actual
+      // operand references differ (as they live in separate modules anyway).
+      if (!std::equal(lhsOp.operand_type_begin(), lhsOp.operand_type_end(),
+                      rhsOp.operand_type_begin())) {
+        return false;
+      }
+
+      // If the operations have regions, recurse into them (depth-first).
+      if (lhsOp.getNumRegions() != rhsOp.getNumRegions()) {
+        return false;
+      }
+      auto lhsRegions = lhsOp.getRegions();
+      auto rhsRegions = rhsOp.getRegions();
+      for (int i = 0; i < lhsRegions.size(); ++i) {
+        if (!areRegionsEquivalent(&lhsRegions[i], &rhsRegions[i])) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool areExecutablesEquivalent(ExecutableOp lhs, ExecutableOp rhs) {
+  auto lhsModule = lhs.getInnerModule();
+  auto rhsModule = rhs.getInnerModule();
+
+  // Must have the same number of entry point ops, with the same attributes.
+  // Entry point op symbol names are expected to differ, that won't affect
+  // equivalence.
+  auto lhsEntryOps = llvm::to_vector<1>(lhsModule.getOps<DispatchEntryOp>());
+  auto rhsEntryOps = llvm::to_vector<1>(rhsModule.getOps<DispatchEntryOp>());
+  if (lhsEntryOps.size() != rhsEntryOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsEntryOps.size(); ++i) {
+    if (lhsEntryOps[i].getAttrs() != rhsEntryOps[i].getAttrs()) {
+      return false;
+    }
+  }
+
+  // Must have the same number of functions, with each listed in the same order
+  // and with equivalent regions inside.
+  auto lhsFuncOps = llvm::to_vector<1>(lhsModule.getOps<FuncOp>());
+  auto rhsFuncOps = llvm::to_vector<1>(rhsModule.getOps<FuncOp>());
+  if (lhsFuncOps.size() != rhsFuncOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsFuncOps.size(); ++i) {
+    auto lhsRegion = lhsFuncOps[i].getCallableRegion();
+    auto rhsRegion = rhsFuncOps[i].getCallableRegion();
+    return areRegionsEquivalent(lhsRegion, rhsRegion);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+class DeduplicateExecutablesPass
+    : public PassWrapper<DeduplicateExecutablesPass, OperationPass<ModuleOp>> {
+ public:
+  explicit DeduplicateExecutablesPass() {}
+  DeduplicateExecutablesPass(const DeduplicateExecutablesPass &pass) {}
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto executableOps = llvm::to_vector<8>(moduleOp.getOps<ExecutableOp>());
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    SmallVector<ExecutableOp, 3> duplicateExecutableOps;
+    DenseMap<Attribute, Attribute> entryPointRefReplacements;
```

Comment:
it's nice when possible to type things as tightly as possible - these could be SymbolRefAttrs for example - as it helps hint to readers what the value is holding and gives you some safety in other places (and no need to cast there)

Reply:
Can they?

I copied this code from https://github.com/google/iree/pull/3470: https://github.com/google/iree/blob/d264a16c0e954c44e9a153f5093b13afc7d0d86b/iree/compiler/Dialect/HAL/Target/VMLA/VMLATarget.cpp#L68-L80

I get build errors after changing from `Attribute` to `SymbolRefAttr`:
```
[build] D:\dev\projects\iree\third_party\llvm-project\llvm\include\llvm/ADT/DenseMap.h(367): error C2039: 'isEqual': is not a member of 'llvm::DenseMapInfo<KeyT>'
[build]         with
[build]         [
[build]             KeyT=mlir::SymbolRefAttr
[build]         ]
```

Reply:
That seems like an oversight, but not one I care too much about. You could at least change the RHS (which is what you cast to above).

Reply:
Changed the RHS.